posthog/temporal/batch_exports/spmc.py

```diff
@@ -474,6 +480,142 @@ def consumer_done_callback(task: asyncio.Task):
     return records_completed
 
 
+Query = str
+QueryParameters = dict[str, typing.Any]
+BatchExportDateRange = tuple[dt.datetime | None, dt.datetime]
+
+
+class RecordBatchModel(abc.ABC):
+    """Base class for models that can be produced as record batches."""
+
+    def __init__(self, team_id: int, is_backfill: bool):
+        self.team_id = team_id
+        self.is_backfill = is_backfill
+
+    async def get_hogql_context(self, team_id: int) -> HogQLContext:
+        """Return a HogQLContext to generate a ClickHouse query."""
+        team = await Team.objects.aget(id=team_id)
+        context = HogQLContext(
+            team=team,
+            team_id=team.id,
+            enable_select_queries=True,
+            limit_top_select=False,
+        )
+        context.database = await database_sync_to_async(create_hogql_database)(team.id, context.modifiers)
+
+        return context
+
+    @abc.abstractmethod
+    async def as_query_with_parameters(
+        self, data_interval_start: dt.datetime | None, data_interval_end: dt.datetime
+    ) -> tuple[Query, QueryParameters]:
+        """Produce a printed query and any necessary ClickHouse query parameters."""
+        raise NotImplementedError
+
+
+class SessionsRecordBatchModel(RecordBatchModel):
+    """A model to produce record batches from the sessions table.
+
+    Attributes:
+       team_id: The ID of the team we are producing records for.
+       is_backfill: Whether we are meant to be backfilling all records.
+    """
+
+    def __init__(self, team_id: int, is_backfill: bool):
+        super().__init__(team_id, is_backfill)
+
+    def get_hogql_query(
+        self, data_interval_start: dt.datetime | None, data_interval_end: dt.datetime
+    ) -> ast.SelectQuery:
+        """Return the HogQLQuery used for the sessions model."""
+        hogql_query = sql.SELECT_FROM_SESSIONS_HOGQL
+
+        where_and = ast.And(
+            exprs=[
+                ast.CompareOperation(
+                    op=ast.CompareOperationOp.Eq,
+                    left=ast.Field(chain=["sessions", "team_id"]),
+                    right=ast.Constant(value=self.team_id),
+                ),
+                ast.CompareOperation(
+                    op=ast.CompareOperationOp.Lt,
+                    left=ast.Field(chain=["_inserted_at"]),
+                    right=ast.Constant(value=data_interval_end),
+                ),
+            ]
+        )
+        if not self.is_backfill and data_interval_start:
```

Comment:
not sure if I'm reading this right but for backfills there will only be an `_inserted_at < {data_interval_end}` clause?

Reply:
Although I would expect `data_interval_start` to be `None` only if `self.is_backfill` is also `True`. But I will simplify here.

Reply:
Simplified, removed all `is_backfill` and left only `not data_interval_start`. If we get no `data_interval_start` from `produce_batch_export_record_batches_from_range` then we omit the filter (which should only happen in backfills).

Reply:
Yeah, backfilling since the beginning of time translates to `data_interval_start=None`.

Reply:
That reminds me I should allow the frontend to also check "Beginning of time" for sessions.

Reply:
(just wondering because I assume this is the/one reason why we don't support this for events?)

Reply:
Hopefully? Plan is to deal with that problem if it happens.